README.md
# About this fork

This project has been forked off and will have bugfixes necessary for continued use in a Celery 3.x cluster.



# Project Status

I've suspended work on this pending a direction on how to handle ensuring singleton tasks across a distributed cluster while keeping the Celery interface intact. It is not enough that the schedulers run -- I need them to either negotiate who runs what (and rebalance on topology changes) or preallocate time periods in Redis to prevent double runs.

As I attempted to use a RedLock and a taint to prevent double runs, I learned that any suspended thread in Celery that blocks long enough for the task to timeout can still cause trouble. When you have >60 workers, you can get disappointment ranging from 3 to 19 issues of "I blocked long enough to screw you".

So the fundamental architecture is at question. We can either plan things out in Time (via preallocating the time segments as a "use once" setup) or in Space (which node thinks it has exclusive control over what and have the other watch it).

# celerybeat-redis

It's modified from celerybeat-mongo (https://github.com/zakird/celerybeat-mongo)

See Changelog in [CHANGES.md](./CHANGES.md)

This is a Celery Beat Scheduler (http://celery.readthedocs.org/en/latest/userguide/periodic-tasks.html)
that stores both the schedules themselves and their status
information in a backend Redis database.

# Features

1. Full-featured celery-beat scheduler
2. Dynamically add/remove/modify tasks
3. Support multiple instance by Active-Standby model

# Installation

It can be installed by
installing the celerybeat-redis Python egg:

    # pip install celerybeat-redis

And specifying the scheduler when running Celery Beat, e.g.

    $ celery beat -S celerybeatredis.schedulers.RedisScheduler

# Configuration

Settings for the scheduler are defined in your celery configuration file
similar to how other aspects of Celery are configured

    CELERY_REDIS_SCHEDULER_URL = "redis://localhost:6379/1"
    CELERY_REDIS_SCHEDULER_KEY_PREFIX = 'tasks:meta:'

You must make sure these two values are configured. `CELERY_REDIS_SCHEDULER_URL`
is used to store tasks. `CELERY_REDIS_SCHEDULER_KEY_PREFIX` is used to generate
keys in redis. The keys will be of the form

    tasks:meta:task-name-here
    tasks:meta:test-fib-every-3s

There is also an optional setting 

    CELERY_REDIS_SCHEDULER_LOCK_TTL = 30

This value determines how long the redis scheduler will hold on to it's lock,
which prevents multiple beat instances from running at the same time. However,
in some cases -- such as a hard crash -- celery beat will not be able to clean
up after itself and release the lock. Therefore, the lock is given a
configurable time-to-live, after which it will expire and be released, if it is
not renewed by the beat instance that acquired it. If said beat instance does
die, another instance will be able to pick up the baton, as it were, and run
instead.

# Quickstart

After installed and configure the needs by above, you can make a try with test, cd to test directory, start a worker by:

    $ celery worker -A tasks -l info

then start the beat by:

    $ celery beat -S celerybeatredis.schedulers.RedisScheduler

celerybeat-redis will load the entry from celeryconfig.py

# Detailed Configuration

There was two ways to add a period task:

## Add in celeryconfig.py

Celery provide a `CELERYBEAT_SCHEDULE` entry in config file, when
celerybeat-redis starts with such a config, it will load tasks to redis, create
them as a celerybeat-redis task.

It's perfect for quick test

## Manaully add to Redis

You can create task by insert specify data to redis, according to following described:

Schedules can be manipulated in the Redis database through
direct database manipulation. There exist two types of schedules,
interval and crontab.

```json
{
    "name" : "interval test schedule",
    "task" : "task-name-goes-here",
    "enabled" : true,
    "interval" : {
        "every" : 5,
        "period" : "minutes"
    },
    "args" : [
        "param1",
        "param2"
    ],
    "kwargs" : {
        "max_targets" : 100
    },
    "total_run_count" : 5,
    "last_run_at" : {
        "__type__": "datetime",
        "year": 2014,
        "month": 8,
        "day": 30,
        "hour": 8,
        "minute": 10,
        "second": 6,
        "microsecond": 667
    }
}
```

The example from Celery User Guide::Periodic Tasks.
```python
CELERYBEAT_SCHEDULE = {
    'interval-test-schedule': {
        'task': 'tasks.add',
        'schedule': timedelta(seconds=30),
        'args': (param1, param2)
    }
}
```

Becomes the following::
```json
{
    "name" : "interval test schedule",
    "task" : "task.add",
    "enabled" : true,
    "interval" : {
        "every" : 30,
        "period" : "seconds",
    },
    "args" : [
        "param1",
        "param2"
    ],
    "kwargs" : {
        "max_targets" : 100
    },
    "total_run_count": 5,
    "last_run_at" : {
        "__type__": "datetime",
        "year": 2014,
        "month": 8,
        "day": 30,
        "hour": 8,
        "minute": 10,
        "second": 6,
        "microsecond": 667
    }
}
```

The following fields are required: name, task, crontab || interval,
enabled when defining new tasks.
`total_run_count` and `last_run_at` are maintained by the
scheduler and should not be externally manipulated.

The example from Celery User Guide::Periodic Tasks.
(see: http://docs.celeryproject.org/en/latest/userguide/periodic-tasks.html#crontab-schedules)

```python
CELERYBEAT_SCHEDULE = {
    # Executes every Monday morning at 7:30 A.M
    'add-every-monday-morning': {
        'task': 'tasks.add',
        'schedule': crontab(hour=7, minute=30, day_of_week=1),
        'args': (16, 16),
    },
}
```

Becomes:

```json
{
    "name" : "add-every-monday-morning",
    "task" : "tasks.add",
    "enabled" : true,
    "crontab" : {
        "minute" : "30",
        "hour" : "7",
        "day_of_week" : "1",
        "day_of_month" : "*",
        "month_of_year" : "*"
    },
    "args" : [
        "16",
        "16"
    ],
    "kwargs" : {},
    "total_run_count" : 1,
    "last_run_at" : {
        "__type__": "datetime",
        "year": 2014,
        "month": 8,
        "day": 30,
        "hour": 8,
        "minute": 10,
        "second": 6,
        "microsecond": 667
    }
}
```

# Deploy multiple nodes

Original celery beat doesn't support multiple node deployment, multiple beat
will send multiple tasks and make worker duplicate execution, celerybeat-redis
use a redis lock to deal with it. Only one node running at a time, other nodes
keep tick with minimal task interval, if this node down, when other node
ticking, it will acquire the lock and continue to run.

WARNING: this is an experiment feature, need more test, not production ready at
this time.


CHANGES.md
v0.0.1 initial release

v0.0.2 Remove task when manually remove task by delete task in redis.

v0.0.6 fix: tick error

v0.0.7 load entries from CELERYBEAT_SCHEDULE thanks @sibson
       Prefix as Parameter thanks @asmodehn

v0.0.8 - v0.1.1 no new feature

v0.1.2 celerybeat-redis can now run in multiple node, only one active, and other nodes are stand by
v0.1.3 support python3
v0.1.5 Refactor for celery 3.1(thanks asmodehn), add ttl for redis lock
v0.1.6 Try for a singleton run of a crontask across the cluster if marked as.

celerybeatredis/task.py
# -*- coding: utf-8 -*-
# Copyright 2014 Kong Luoxing

# Licensed under the Apache License, Version 2.0 (the 'License'); you may not
# use this file except in compliance with the License. You may obtain a copy
# of the License at http://www.apache.org/licenses/LICENSE-2.0
import datetime
import functools
import hashlib
from copy import deepcopy
import celery
import redis.exceptions
import json

from .decoder import DateTimeDecoder, DateTimeEncoder
from .exceptions import TaskTypeError
from .globals import bytes_to_str, default_encoding, logger


def catch_errors(func):
    @functools.wraps(func)
    def wrapper(*args, **kwargs):
        try:
            return func(*args, **kwargs)
        except Exception:
            logger.exception('Unhandled EXC in {}'.format(func.__name__))
            raise
    return wrapper


class Interval(object):

    def __init__(self, every, period='seconds'):
        self.every = every
        # could be seconds minutes hours
        self.period = period

    @property
    def period_singular(self):
        return self.period[:-1]

    def __unicode__(self):
        if self.every == 1:
            return 'every {0.period_singular}'.format(self)
        return 'every {0.every} {0.period}'.format(self)


class Crontab(object):

    def __init__(self, minute=0, hour=0, day_of_week=None, day_of_month=None, month_of_year=None):
        self.minute = minute
        self.hour = hour
        self.day_of_week = day_of_week or '*'
        self.day_of_month = day_of_month or '*'
        self.month_of_year = month_of_year or '*'

    def __unicode__(self):
        rfield = lambda f: f and str(f).replace(' ', '') or '*'
        return '{0} {1} {2} {3} {4} (m/h/d/dM/MY)'.format(
            rfield(self.minute), rfield(self.hour), rfield(self.day_of_week),
            rfield(self.day_of_month), rfield(self.month_of_year),
        )


class PeriodicTask(object):
    """
    Represents a periodic task.
    This follows the celery.beat.ScheduleEntry class design.
    However it is independent of any celery import, so that any client library can import this module
     and use it to manipulate periodic tasks into a Redis database, without worrying about all the celery imports.
    Should follow the SQLAlchemy DBModel design.
    These are used as delegate from https://github.com/celery/django-celery/blob/master/djcelery/schedulers.py
    """
    name = None
    task = None

    data = None

    args = []
    kwargs = {}
    options = {}

    enabled = True

    # datetime
    last_run_at = None
    last_task_id = None

    total_run_count = 0

    # Follow celery.beat.SchedulerEntry:__init__() signature as much as possible
    def __init__(self, name, task, schedule, key=None, enabled=True, args=(), kwargs=None, options=None,
                 last_run_at=None, last_task_id=None, total_run_count=None, **extrakwargs):
        """
        :param name: name of the task ( = redis key )
        :param task: taskname ( as in celery : python function name )
        :param schedule: the schedule. maybe also a dict with all schedule content
        :param relative: if the schedule time needs to be relative to the interval ( see celery.schedules )
        :param enabled: whether this task is enabled or not
        :param args: args for the task
        :param kwargs: kwargs for the task
        :param options: options for hte task
        :param last_run_at: lat time the task was run
        :param last_task_id: the last id otf the task triggered by this periodic task
        :param total_run_count: total number of times the task was run
        :return:
        """

        self.task = task
        self.enabled = enabled

        # Using schedule property conversion
        # logger.warn("Schedule in Task init {s}".format(s=schedule))
        self.schedule = schedule

        self.args = args
        self.kwargs = kwargs or {}
        self.options = options or {}

        self.last_run_at = last_run_at
        self.last_task_id = last_task_id
        self.total_run_count = total_run_count

        self.name = name
        self.key = key if key else self.name
        self.delete_key = 'deleted:' + bytes_to_str(self.key)

        self.running = False
        self.singleton = kwargs.get('singleton', False)

        # storing extra arguments (might be useful to have other args depending on application)
        for elem in extrakwargs.keys():
            setattr(self, elem, extrakwargs[elem])

    @staticmethod
    @catch_errors
    def get_all_as_dict(rdb, key_prefix):
        """get all of the tasks, for best performance with large amount of tasks, return a generator
        """

        tasks = rdb.keys(key_prefix + '*')
        for task_key in tasks:
            try:
                value, = rdb.hmget(task_key, 'schedule')
                logger.debug('Loading {}'.format(task_key))
                if value is None:
                    logger.debug('{} was erased!'.format(task_key))
                    continue
                dct = json.loads(
                    bytes_to_str(value),
                    cls=DateTimeDecoder, encoding=default_encoding)
                # task name should always correspond to the key in redis to avoid
                # issues arising when saving keys - we want to add information to
                # the current key, not create a new key
                # logger.warning('json task {0}'.format(dct))
                yield task_key, dct
            except redis.exceptions.ResponseError as e:
                if str(e).startswith('WRONGTYPE'):
                    logger.error(
                        '{} contains incorrect data structure. Destroying.'.format(task_key))
                    rdb.delete(task_key)
                    continue
                raise
            except ValueError:
                logger.exception('Unable to decode data at {}. Possibly deleted'.format(task_key))

    def _next_instance(self, last_run_at):
        self.last_run_at = last_run_at
        self.total_run_count += 1
        """Return a new instance of the same class, but with
        its date and count fields updated."""
        return self.__class__(**dict(
            self,
            last_run_at=last_run_at,
            total_run_count=self.total_run_count + 1,
        ))
    __next__ = next = _next_instance  # for 2to3

    @catch_errors
    def jsondump(self):
        # must do a deepcopy using our custom iterator to choose what to save # BTJ: Why?
        # (matching external view)
        self_dict = deepcopy({k: v for k, v in iter(self) if v is not None})
        logger.debug('Json dump! {}'.format(self_dict))
        return json.dumps(self_dict, cls=DateTimeEncoder)

    @catch_errors
    def jsonhash(self):
        hashed = hashlib.sha1()
        for key, value in self:
            hashed.update(key)
            hashed.update(str(value))
        return hashed.hexdigest()

    @catch_errors
    def update(self, other):
        """
        Update values from another task.
        This is used to dynamically update periodic task from edited redis values
        Does not update "non-editable" fields (last_run_at, total_run_count).
        Extra arguments will be updated (considered editable)
        """
        otherdict = other.__dict__  # note : schedule property is not part of the dict.
        otherdict.pop('last_run_at')
        otherdict.pop('last_task_id')
        otherdict.pop('total_run_count')
        self.__dict__.update(otherdict)

    def __repr__(self):
        return '<PeriodicTask ({0} {1}(*{2}, **{3}) options: {4} schedule: {5})>'.format(
            self.name, self.task, self.args,
            self.kwargs, self.options, self.schedule,
        )

    def __unicode__(self):
        fmt = '{0.name}: {0.schedule}'
        return fmt.format(self)

    @catch_errors
    def get_schedule(self):
        """
        schedule Interval / Crontab -> dict
        :return:
        """
        return vars(self.data)

    @catch_errors
    def set_schedule(self, schedule):
        """
        schedule dict -> Interval / Crontab if needed
        :return:
        """
        if schedule is None:
            pass
        elif isinstance(schedule, Interval) or isinstance(schedule, Crontab):
            self.data = schedule
        elif isinstance(schedule, celery.schedules.crontab):
            self.data = Crontab(schedule.minute, schedule.hour,
                                schedule.day_of_week, schedule.day_of_month, schedule.month_of_year)
        elif isinstance(schedule, celery.schedules.schedule)\
                or isinstance(schedule, datetime.timedelta):
            self.data = Interval(schedule.seconds)
        else:
            schedule_inst = None
            for s in (Interval, Crontab):
                try:
                    schedule_inst = s(**schedule)
                except TypeError:
                    pass
                    # logger.warn("Create schedule failed. {}".format(schedule.__class__))

            if schedule_inst is None:
                raise TaskTypeError(
                    "Schedule {s} didn't match Crontab or Interval type".format(s=schedule))
            else:
                self.data = schedule_inst

    schedule = property(get_schedule, set_schedule)

    @catch_errors
    def __iter__(self):
        """
        We iterate on our members a little bit specially
        => data is hidden and schedule is shown instead
        => rdb is hidden
        :return:
        """
        data = dict(vars(self))
        if 'data' in data:
            del data['data']
            data['schedule'] = self.schedule
        keys = sorted(data)
        for key in keys:
            yield key, data[key]


celerybeatredis/schedulers.py
# -*- coding: utf-8 -*-
# Copyright 2014 Kong Luoxing

# Licensed under the Apache License, Version 2.0 (the 'License'); you may not
# use this file except in compliance with the License. You may obtain a copy
# of the License at http://www.apache.org/licenses/LICENSE-2.0
import contextlib
import datetime
import functools
import logging
import threading
import time

import celery.schedules
import kombu.utils
import redis.exceptions
from celery import current_app
from celery.beat import Scheduler
from celery.result import ResultBase
from redis import StrictRedis
from redlock import Redlock, MultipleRedlockException, Lock

from .exceptions import TaskTypeError
from .task import PeriodicTask, catch_errors


logger = logging.getLogger(__name__)

# we don't need simplejson, builtin json module is good enough


class EmptyResult(ResultBase):
    '''
    Done to convince celery to not freak out
    '''
    task = 'Not a task'
    id = 'faked id'


@catch_errors
def lock_task_until(name, dlm, lock, t_s, db, result, after_func):
    logger.info('Starting singleton task thread tracking {!r} using a redis lock {} ({})'.format(
        name, lock.resource, lock.key))
    try:
        while not result.ready():
            logger.info('Refresh token {!r}'.format(lock.resource))
            dlm.touch(lock, 4*60*1000)
            time.sleep(4*60)

    except Exception:
        logger.exception('Error in {}'.format(name))
    finally:
        logger.info('Callback on {}. Took {:.2f}s'.format(name, time.time() - t_s))
        try:
            logger.info('Releasing Task {} lock'.format(name))
            dlm.unlock(lock)
            after_func()
        except Exception:
            logger.exception('Unable to release lock on behalf of {!r}'.format(name))

    try:
        result.get()
    except Exception:
        logger.exception('Error in {}'.format(name))


class RedisScheduleEntry(object):
    """
    The Schedule Entry class is mainly here to handle the celery dependency injection
     and delegates everything to a PeriodicTask instance
    It follows the Adapter Design pattern, trivially implemented in python with __getattr__
        and __setattr__
    This is similar to https://github.com/celery/django-celery/blob/master/djcelery/schedulers.py
     that uses SQLAlchemy DBModels as delegates.
    """
    rdb = None

    def __init__(self, name=None, task=None, enabled=True, last_run_at=None,
                 total_run_count=None, schedule=None, args=(), kwargs=None,
                 options=None, app=None, **extrakwargs):

        # defaults (MUST NOT call self here - or loop __getattr__ for ever)
        app = app or current_app
        # Setting a default time a bit before now to not miss a task that was just added.
        last_run_at = last_run_at or app.now() - datetime.timedelta(
                seconds=app.conf.CELERYBEAT_MAX_LOOP_INTERVAL)

        # using periodic task as delegate
        self._task = PeriodicTask(
                # Note : for compatibiilty with celery methods, the name of the task is
                # actually the key in redis DB.
                # For extra fancy fields (like a human readable name, you can leverage extrakwargs)
                name=name,
                task=task,
                enabled=enabled,
                schedule=schedule,  # TODO : sensible default here ?
                args=args,
                kwargs=kwargs or {},
                options=options or {},
                last_run_at=last_run_at,
                total_run_count=total_run_count or 0,
                **extrakwargs
        )

        self.app = app
        super(RedisScheduleEntry, self).__init__()

    # automatic delegation to PeriodicTask (easy delegate)
    def __getattr__(self, attr):
        return getattr(self._task, attr)

    def __setattr__(self, attr, value):
        if attr in ('app', '_task', 'rdb', 'singleton'):
            return super(RedisScheduleEntry, self).__setattr__(attr, value)
        # We set the attribute in the task delegate if available
        if self._task and hasattr(self._task, attr):
            return setattr(self._task, attr, value)

        # else we raise
        raise AttributeError(
                "Attribute {attr} not found in {tasktype}".format(attr=attr,
                                                                  tasktype=type(self._task)))

    #
    # Overrides schedule accessors in PeriodicTask to store dict in json but
    # retrieve proper celery schedules
    #
    def get_schedule(self):
        if {'every', 'period'}.issubset(self._task.schedule.keys()):
            return celery.schedules.schedule(
                    datetime.timedelta(
                        **{self._task.schedule['period']: self._task.schedule['every']}),
                    self.app)
        elif {'minute', 'hour', 'day_of_week', 'day_of_month', 'month_of_year'}.issubset(
                self._task.schedule.keys()):
            return celery.schedules.crontab(minute=self._task.schedule['minute'],
                                            hour=self._task.schedule['hour'],
                                            day_of_week=self._task.schedule['day_of_week'],
                                            day_of_month=self._task.schedule['day_of_month'],
                                            month_of_year=self._task.schedule['month_of_year'],
                                            app=self.app)
        else:
            raise TaskTypeError('Existing Task schedule type not recognized')

    def set_schedule(self, schedule):
        if isinstance(schedule, celery.schedules.schedule):
            # TODO : unify this with Interval in PeriodicTask
            self._task.schedule = {
                'every': max(schedule.run_every.total_seconds(), 0),
                'period': 'seconds'
            }
        elif isinstance(schedule, celery.schedules.crontab):
            # TODO : unify this with Crontab in PeriodicTask
            self._task.schedule = {
                'minute': schedule._orig_minute,
                'hour': schedule._orig_hour,
                'day_of_week': schedule._orig_day_of_week,
                'day_of_month': schedule._orig_day_of_month,
                'month_of_year': schedule._orig_month_of_year
            }
        else:
            raise TaskTypeError('New Task schedule type not recognized')

    schedule = property(get_schedule, set_schedule)

    #
    # Overloading ScheduleEntry methods
    #
    def is_due(self):
        """See :meth:`~celery.schedule.schedule.is_due`."""
        due = self.schedule.is_due(self.last_run_at)

        logger.debug('task {0} due : {1}'.format(self.name, due))
        if not self.enabled:
            logger.info('task {0} is disabled. not triggered.'.format(self.name))
            # if the task is disabled, we always return false, but the time that
            # it is next due is returned as usual
            return celery.schedules.schedstate(is_due=False, next=due[1])
        return due

    def __repr__(self):
        return '<RedisScheduleEntry: {0.name} {call} {0.schedule}'.format(
                self,
                call=kombu.utils.reprcall(self.task, self.args or (), self.kwargs or {}),
        )

    def update(self, other):
        """
        Update values from another entry.
        This is used to dynamically update periodic entry from edited redis values
        Does not update "non-editable" fields
        Extra arguments will be updated (considered editable)
        """
        # Handle delegation properly here
        self._task.update(other._task)
        # we should never need to touch the app here

    #
    # ScheduleEntry needs to be an iterable
    #

    # from celery.beat.ScheduleEntry._default_now
    def _default_now(self):
        return self.get_schedule().now() if self.schedule else self.app.now()

    # from celery.beat.ScheduleEntry._next_instance
    def _next_instance(self, last_run_at=None):
        """Return a new instance of the same class, but with
        its date and count fields updated."""
        return self.__class__(**dict(
                self,
                last_run_at=last_run_at or self._default_now(),
                total_run_count=self.total_run_count + 1,
        ))

    __next__ = next = _next_instance  # for 2to3

    def __iter__(self):
        # We need to delegate iter (iterate on task members, not on multiple tasks)
        # Following celery.SchedulerEntry.__iter__() design
        return iter(self._task)

    @staticmethod
    def get_all_as_dict(scheduler_url, key_prefix):
        """get all of the tasks, for best performance with large amount of tasks, return a generator
        """
        # Calling another generator
        for task_key, task_dict in PeriodicTask.get_all_as_dict(scheduler_url, key_prefix):
            yield task_key, task_dict

    @classmethod
    def from_entry(cls, scheduler_url, name, **entry):
        # options = entry.get('options') or {}  # unused variable!
        fields = dict(entry)
        fields['name'] = current_app.conf.CELERY_REDIS_SCHEDULER_KEY_PREFIX + name
        schedule = fields.pop('schedule')
        schedule = celery.schedules.maybe_schedule(schedule)
        if isinstance(schedule, celery.schedules.crontab):
            fields['crontab'] = {
                'minute': schedule._orig_minute,
                'hour': schedule._orig_hour,
                'day_of_week': schedule._orig_day_of_week,
                'day_of_month': schedule._orig_day_of_month,
                'month_of_year': schedule._orig_month_of_year
            }
        elif isinstance(schedule, celery.schedules.schedule):
            fields['interval'] = {'every': max(schedule.run_every.total_seconds(), 0),
                                  'period': 'seconds'}

        fields['args'] = fields.get('args', [])
        fields['kwargs'] = fields.get('kwargs', {})
        fields['key'] = fields['name']
        return cls(PeriodicTask.from_dict(fields, scheduler_url))


def lock_factory(dlm, name, ttl_s):

    @contextlib.contextmanager
    def lock():
        locked = False
        try:
            logger.info('Secure lock for {} with a ttl of {}'.format(name, ttl_s))
            locked = dlm.lock(name, ttl_s*1000)
            if not isinstance(locked, Lock):
                raise redis.exceptions.LockError('Unable to lock')
            yield locked
        except MultipleRedlockException as e:
            raise redis.exceptions.LockError(e)
        finally:
            if locked is not False:
                logger.info('Releasing {}'.format(locked.resource))
                dlm.unlock(locked)
    return lock

DEFAULT_REDIS_URI = 'redis://localhost'


class RedisScheduler(Scheduler):
    Entry = RedisScheduleEntry

    def __init__(self, *args, **kwargs):
        if hasattr(current_app.conf, 'CELERY_REDIS_SCHEDULER_URL'):
            logger.info('backend scheduler using %s',
                        current_app.conf.CELERY_REDIS_SCHEDULER_URL)
        else:
            logger.info('backend scheduler using %s', DEFAULT_REDIS_URI)

        self.update_interval = current_app.conf.get('UPDATE_INTERVAL') or datetime.timedelta(
                seconds=10)

        # how long we should hold on to the redis lock in seconds
        if 'CELERY_REDIS_SCHEDULER_LOCK_TTL' in current_app.conf:
            lock_ttl = current_app.conf.CELERY_REDIS_SCHEDULER_LOCK_TTL
        else:
            lock_ttl = 30

        if lock_ttl < self.update_interval.seconds:
            lock_ttl = self.update_interval.seconds * 2
        self.lock_ttl = lock_ttl

        self._dirty = set()  # keeping modified entries by name for sync later on
        self._schedule = {}  # keeping dynamic schedule from redis DB here
        # self.data is used for statically configured schedule
        try:
            self.schedule_url = current_app.conf.CELERY_REDIS_SCHEDULER_URL
        except AttributeError:
            self.schedule_url = DEFAULT_REDIS_URI

        self.rdb = StrictRedis.from_url(self.schedule_url)
        logger.info('Setting RedLock provider to {}'.format(self.schedule_url))
        self.dlm = Redlock([self.rdb])
        self._secure_cronlock = \
            lock_factory(self.dlm, 'celery:beat:task_lock',  self.lock_ttl)
        self._last_updated = None

        self.Entry.scheduler = self
        self.Entry.rdb = self.rdb

        # This will launch setup_schedule if not lazy
        super(RedisScheduler, self).__init__(*args, **kwargs)
        logger.info('Scheduler ready')

    @catch_errors
    def setup_schedule(self):
        super(RedisScheduler, self).setup_schedule()
        # In case we have a preconfigured schedule
        # self.update_from_dict(self.app.conf.CELERYBEAT_SCHEDULE)

        prefix = current_app.conf.CELERY_REDIS_SCHEDULER_KEY_PREFIX
        signature = None
        deferreds = []

        for name, entry in self.app.conf.CELERYBEAT_SCHEDULE.items():
            entry = self._maybe_entry(name, entry)
            logger.debug('{} -> {}'.format(name, entry.jsondump()))

            schedule_hash = entry.jsonhash()
            key = name
            if not name.startswith(prefix):
                key = '{}{}'.format(prefix, name)

            try:
                signature = self.rdb.hget(key, 'hash')
            except redis.exceptions.ResponseError:
                logger.exception('Fetching {} threw an error. Clearing it.'.format(key))
                signature = None

            if signature != schedule_hash:
                self.rdb.delete(key)
                deferreds.append((key, {
                    'hash': schedule_hash,
                    'schedule': entry.jsondump()
                    }))

        if deferreds:
            try:
                with self._secure_cronlock() as lock:
                    t_s = time.time()

                    if not lock:
                        logger.debug('Unable to acquire write lock')
                        return
                    for key, value in deferreds:
                        if time.time() - t_s >= 30:
                            self.dlm.touch(lock, 60*1000)
                            t_s = time.time()
                        logger.debug('Update/insert {} into Redis -> {}'.format(key, value))
                        self.rdb.hmset(key, value)

            except redis.exceptions.LockError:
                logger.exception('Unable to acquire write lock, encountered some redis errors')
            else:
                try:
                    s = self.all_as_schedule()
                    self.merge_inplace(s)
                except Exception:
                    logger.exception('Unable to merge in place')
        self.update_from_dict(self.app.conf.CELERYBEAT_SCHEDULE)

    @catch_errors
    def tick(self):
        """Run a tick, that is one iteration of the scheduler.
        Executes all due tasks.
        """
        # need to grab all data (might have been updated) from schedule DB.
        # we need to merge it with whatever schedule was set in config, and already
        # installed default tasks
        try:
            s = self.all_as_schedule()
            self.merge_inplace(s)
        except Exception as exc:
            logger.error(
                    "Exception when getting tasks from {url} : {exc}".format(url=self.schedule_url,
                                                                             exc=exc))
            # TODO : atomic merge : be able to cancel it if there s a problem
            raise

        # # displaying the schedule we got from redis
        logger.debug("DB schedule : {0}".format(self.schedule))

        # this will call self.maybe_due() to check if any entry is due.
        return super(RedisScheduler, self).tick()

    @catch_errors
    def all_as_schedule(self, key_prefix=None, entry_class=None):
        logger.debug('RedisScheduler: Fetching database schedule')
        key_prefix = key_prefix or current_app.conf.CELERY_REDIS_SCHEDULER_KEY_PREFIX
        entry_class = entry_class or self.Entry

        d = {}
        for key, task in entry_class.get_all_as_dict(self.rdb, key_prefix):
            # logger.debug('Building {0} from : {1}'.format(entry_class, task))
            d[key] = entry_class(**dict(task, app=self.app))
        return d

    @catch_errors
    def reserve(self, entry):
        # called when the task is about to be run
        # (and data will be modified -> sync() will need to save it)

        new_entry = super(RedisScheduler, self).reserve(entry)
        # Need to store the key of the entry, because the entry may change in the mean time.
        self._dirty.add(new_entry.name)
        return new_entry

    def apply_async(self, entry, publisher=None, **kwargs):
        singleton = False

        if isinstance(entry, RedisScheduleEntry):
            singleton = entry.singleton
        if not singleton:
            logger.debug('Not a singleton')
            return super(RedisScheduler, self).apply_async(
                entry, entry, publisher=publisher, **kwargs)

        logger.info('Attempting to secure an exclusive lock for {}'.format(entry))
        lock_name = '{}:run'.format(entry.name)

        if self.rdb.get(lock_name + ':taint'):
            logger.warning('Taint found for {}'.format(entry.name))
            return EmptyResult()

        ctx_manager = lock_factory(self.dlm, lock_name,  5*60)()
        try:
            lock = ctx_manager.__enter__()
            if lock is False:
                logger.debug('Unable to secure {}'.format(entry.name))
                return EmptyResult()

        except redis.exceptions.LockError:
            logger.debug('Unable to secure {}'.format(entry.name))
            return EmptyResult()

        except Exception:
            logger.exception('Error in lock secure')
            return EmptyResult()

        logger.info('Running {}'.format(entry.name))
        t_s = time.time()
        ttl = self.rdb.ttl(lock.resource)

        assert self.rdb.get(lock.resource), 'Lock {} never materialized'.format(lock.resource)
        assert ttl > 120, 'Lock {} never materialized with a bad ttl {}'.format(lock.resource, ttl)
        result = super(RedisScheduler, self).apply_async(
            entry, producer=publisher, **kwargs)

        logger.info('Attaching callback thread for {}'.format(entry.name))
        t = threading.Thread(
            target=lock_task_until, args=(
                entry.name, self.dlm, lock, t_s, self.rdb, result,
                lambda: self.rdb.set(lock_name+':taint', t_s, ex=120)))
        t.daemon = True
        t.start()

        return result

    @catch_errors
    def close(self):
        self.sync()

    @property
    def info(self):
        return '    . db -> {self.schedule_url}'.format(self=self)
